File: backend/src/pipeline/evaluator.ts
/**
 * Final evaluation. The LLM scores five rubric dimensions (0-100) and returns
 * concise strengths/gaps; WE compute the weighted overall score and the
 * recommendation here so they stay consistent with the displayed breakdown.
 */
import { z } from "zod";
import { config } from "../config.js";
import { buildEvalPrompt, EVAL_SCHEMA } from "../llm/prompts.js";
import { htmlToText } from "../llm/htmlToText.js";
import type { GeminiCaller } from "./submodels.js";
import type { Candidate } from "./candidate.js";

const DimensionScore = z.object({
  score: z.number().int().min(0).max(100),
  reason: z.string(),
});

const EvaluationLlm = z.object({
  dimensions: z.object({
    hard_skills: DimensionScore,
    experience_relevance: DimensionScore,
    seniority_scope: DimensionScore,
    education_certs: DimensionScore,
    domain_knowledge: DimensionScore,
  }),
  strengths: z.array(z.string()),
  gaps: z.array(z.string()),
});

export type DimensionKey = keyof z.infer<typeof EvaluationLlm>["dimensions"];

export interface EvaluationResult {
  matchScore: number;
  recommendation: "strong_match" | "good_match" | "reject";
  dimensions: Record<DimensionKey, { score: number; weight: number; reason: string }>;
  strengths: string[];
  gaps: string[];
}

/** Trim each bullet to the word cap and limit how many we keep. */
function trimBullets(items: string[]): string[] {
  const { maxBullets, maxBulletWords } = config.rubric;
  return items
    .map((s) => s.trim())
    .filter(Boolean)
    .slice(0, maxBullets)
    .map((s) => {
      const words = s.split(/\s+/);
      return words.length <= maxBulletWords ? s : words.slice(0, maxBulletWords).join(" ") + "…";
    });
}

/** Cheap deterministic signals to ground the LLM's hard-skills judgement. */
function computeSignals(jdText: string, candidate: Candidate) {
  const jd = jdText.toLowerCase();
  const matchedSkills = candidate.skills.filter((s) => jd.includes(s.toLowerCase()));
  return {
    matchedSkills,
    skillMatchCount: matchedSkills.length,
    totalSkills: candidate.skills.length,
    experienceYears: candidate.experienceYears,
  };
}

/** Compact candidate view sent to the model (drops bulky link metadata). */
function candidateForPrompt(candidate: Candidate) {
  return {
    name: candidate.name,
    skills: candidate.skills,
    experienceYears: candidate.experienceYears,
    experience: candidate.experienceText,
    education: candidate.education,
    certifications: candidate.certifications,
    linkCategories: [...new Set(candidate.links.map((l) => l.category))],
  };
}

function finalize(parsed: z.infer<typeof EvaluationLlm>): EvaluationResult {
  const { weights, thresholds } = config.rubric;
  const dimensions = {} as EvaluationResult["dimensions"];
  let total = 0;
  for (const key of Object.keys(weights) as DimensionKey[]) {
    const weight = weights[key];
    const { score, reason } = parsed.dimensions[key];
    dimensions[key] = { score, weight, reason };
    total += score * weight;
  }
  const matchScore = Math.round(total);
  const recommendation =
    matchScore >= thresholds.strong
      ? "strong_match"
      : matchScore >= thresholds.good
        ? "good_match"
        : "reject";

  return {
    matchScore,
    recommendation,
    dimensions,
    strengths: trimBullets(parsed.strengths),
    gaps: trimBullets(parsed.gaps),
  };
}

export async function evaluate(
  jobDescription: string,
  candidate: Candidate,
  call: GeminiCaller,
): Promise<EvaluationResult> {
  // The JD may be rich-text HTML; feed the evaluator clean plain text.
  const jdText = htmlToText(jobDescription);
  const signals = computeSignals(jdText, candidate);
  const res = await call({
    prompt: buildEvalPrompt(
      jdText,
      candidateForPrompt(candidate),
      signals,
      config.rubric.maxBulletWords,
      config.rubric.maxBullets,
    ),
    schema: EVAL_SCHEMA,
  });
  return finalize(EvaluationLlm.parse(res));
}

File: backend-py/app/pipeline/evaluator.py
"""Final evaluation (port of evaluator.ts).

The LLM scores five rubric dimensions (0-100) and returns concise strengths/gaps;
WE compute the weighted overall score and the recommendation here so they stay
consistent with the displayed breakdown.
"""

from __future__ import annotations

import math
from typing import Any, Awaitable, Callable

from app.config import (
    MAX_BULLET_WORDS,
    MAX_BULLETS,
    RUBRIC_THRESHOLDS,
    RUBRIC_WEIGHTS,
)
from app.llm.gemini import GeminiRequest
from app.llm.html_to_text import html_to_text
from app.llm.prompts import EVAL_SCHEMA, build_eval_prompt

GeminiCaller = Callable[[GeminiRequest], Awaitable[Any]]


def _trim_bullets(items: list[str]) -> list[str]:
    """Trim each bullet to the word cap and limit how many we keep."""
    cleaned = [s.strip() for s in items]
    cleaned = [s for s in cleaned if s][:MAX_BULLETS]
    out = []
    for s in cleaned:
        words = s.split()
        if len(words) <= MAX_BULLET_WORDS:
            out.append(s)
        else:
            out.append(" ".join(words[:MAX_BULLET_WORDS]) + "…")
    return out


def _compute_signals(jd_text: str, candidate: dict) -> dict:
    """Cheap deterministic signals to ground the LLM's hard-skills judgement."""
    jd = jd_text.lower()
    skills = candidate.get("skills") or []
    matched = [s for s in skills if s.lower() in jd]
    return {
        "matchedSkills": matched,
        "skillMatchCount": len(matched),
        "totalSkills": len(skills),
        "experienceYears": candidate.get("experienceYears"),
    }


def _candidate_for_prompt(candidate: dict) -> dict:
    """Compact candidate view sent to the model (drops bulky link metadata)."""
    links = candidate.get("links") or []
    link_categories = list(dict.fromkeys(l.get("category") for l in links))
    return {
        "name": candidate.get("name"),
        "skills": candidate.get("skills") or [],
        "experienceYears": candidate.get("experienceYears"),
        "experience": candidate.get("experienceText"),
        "education": candidate.get("education") or [],
        "certifications": candidate.get("certifications") or [],
        "linkCategories": link_categories,
    }


def _finalize(parsed: dict) -> dict:
    dimensions: dict[str, Any] = {}
    total = 0.0
    for key, weight in RUBRIC_WEIGHTS.items():
        dim = parsed["dimensions"][key]
        score = int(dim["score"])
        dimensions[key] = {"score": score, "weight": weight, "reason": dim["reason"]}
        total += score * weight

    # Match JS Math.round (round half up), not Python's banker's rounding.
    match_score = math.floor(total + 0.5)
    if match_score >= RUBRIC_THRESHOLDS["strong"]:
        recommendation = "strong_match"
    elif match_score >= RUBRIC_THRESHOLDS["good"]:
        recommendation = "good_match"
    else:
        recommendation = "reject"

    return {
        "matchScore": match_score,
        "recommendation": recommendation,
        "dimensions": dimensions,
        "strengths": _trim_bullets(parsed.get("strengths") or []),
        "gaps": _trim_bullets(parsed.get("gaps") or []),
    }


async def evaluate(job_description: str, candidate: dict, call: GeminiCaller) -> dict:
    # The JD may be rich-text HTML; feed the evaluator clean plain text.
    jd_text = html_to_text(job_description)
    signals = _compute_signals(jd_text, candidate)
    res = await call(
        GeminiRequest(
            prompt=build_eval_prompt(
                jd_text,
                _candidate_for_prompt(candidate),
                signals,
                MAX_BULLET_WORDS,
                MAX_BULLETS,
            ),
            schema=EVAL_SCHEMA,
        )
    )
    return _finalize(res)
